### PR TITLE
BUG: computing centreline using physical coordinates instead of voxel…

### LINKED ISSUE #1188: NURBS are unstable when calculated in voxel coordinates
especially when voxel spacing in one direction is very large.

This instability causes fluctuations of the derivatives, which induces wrong calculations of the angles. This has an strong effect on CSA computation, as demonstrated in the example below.

Data: `sct_issues/20170301_issue1188/`

Command:
```
sct_process_segmentation -i t2s_segm.nii.gz -p csa
```

Results:
```
# Slice (z),CSA (mm^2),Angle with respect to the I-S direction (degrees)
0,63.5923927147,18.6837707469
1,79.2658753014,8.26795055912
2,88.2874975915,5.12747654869
3,87.9995780207,6.03282331933
4,-8.61862123509,95.2875209491
5,89.2278470889,4.91847828944
6,95.8647986243,9.45846253778
7,84.0700371176,4.83530051485
8,84.7877521872,1.7651865876
9,81.1628827015,0.530803473898
10,82.0784541216,0.515470049064
11,80.383388821,1.2822922865
12,83.1066671944,1.84459163675
```

### PR DESCRIPTION
**Pull request by benjamindeleener**
… coordinates. Fixed #1188

### CODE REVIEW

File: scripts/sct_process_segmentation.py [hunk 1]
```diff
@@ -750,7 +767,7 @@ def compute_csa(fname_segmentation, output_folder, overwrite, verbose, remove_te
     for i in range(min_z_index, max_z_index+1):
         file_results.write(str(int(i)) + ',' + str(csa[i-min_z_index])+ ',' + str(angles[i-min_z_index])+'\n')
         # Display results
-        sct.printv('z = '+str(i-min_z_index)+', CSA = '+str(csa[i-min_z_index])+' mm^2'+', Angle = '+str(angles[i-min_z_index])+' deg', type='info')
+        sct.printv('z = '+str(i)+', CSA = '+str(csa[i-min_z_index])+' mm^2'+', Angle = '+str(angles[i-min_z_index])+' deg', type='info')
```
Comment (peristeri) [inline scripts/sct_process_segmentation.py]: `'z = %d, CSA = %f mm^2, Angle = %f deg' % (i, csa[i-min_z_index], angles[i-min_z_index])` is a easier to read.

File: scripts/sct_process_segmentation.py [hunk 2]
```diff
@@ -625,11 +626,27 @@ def compute_csa(fname_segmentation, output_folder, overwrite, verbose, remove_te
     X, Y, Z = (data_seg > 0).nonzero()
     min_z_index, max_z_index = min(Z), max(Z)
 
-    # fit centerline, smooth it and return the first derivative (in voxel space but FITTED coordinates)
-    x_centerline_fit, y_centerline_fit, z_centerline, x_centerline_deriv, y_centerline_deriv, z_centerline_deriv = smooth_centerline('segmentation_RPI.nii.gz', algo_fitting=algo_fitting, type_window=type_window, window_length=window_length, nurbs_pts_number=3000, phys_coordinates=False, verbose=verbose, all_slices=True)
+    use_phys_coord = True
```
Comment (peristeri) [inline scripts/sct_process_segmentation.py]: Do you need `use_phys_coord` variable?

File: scripts/sct_process_segmentation.py [hunk 3]
```diff
@@ -625,11 +626,27 @@ def compute_csa(fname_segmentation, output_folder, overwrite, verbose, remove_te
...
+        x_centerline_fit, y_centerline_fit, z_centerline, x_centerline_deriv, y_centerline_deriv, z_centerline_deriv = smooth_centerline('segmentation_RPI.nii.gz', algo_fitting=algo_fitting, type_window=type_window, window_length=window_length, nurbs_pts_number=3000, phys_coordinates=True, verbose=verbose, all_slices=False)
+        centerline = Centerline(x_centerline_fit, y_centerline_fit, z_centerline, x_centerline_deriv, y_centerline_deriv, z_centerline_deriv)
+
+        # average centerline coordinates over slices of the image
+        x_centerline_fit_rescorr, y_centerline_fit_rescorr, z_centerline_rescorr, x_centerline_deriv_rescorr, y_centerline_deriv_rescorr, z_centerline_deriv_rescorr = centerline.average_coordinates_over_slices(im_seg)
+
+        # compute Z axis of the image, in physical coordinate
+        axis_X, axis_Y, axis_Z = im_seg.get_directions()
+
+    else:
```
Comment (peristeri) [inline scripts/sct_process_segmentation.py]: Since the variable is hard coded as `True`, you don't need the else code.

Comment (benjamindeleener) [inline scripts/sct_process_segmentation.py]: Oops I forgot to make it available. Nice catch! I'll change it.

File: scripts/sct_process_segmentation.py (hunk 1 above)
Comment (benjamindeleener) [inline scripts/sct_process_segmentation.py]: That's a point of view ;-) I'll change

**Pull request merged**